pkg/controllers/state/cost.go

```diff
@@ -0,0 +1,370 @@
+	npCost, exists := cc.npCostMap[np.Name]
+
+	if !exists {
+		cc.createNewNodePoolCost(np, instanceTypes)
+	} else {
+		npCost.overlayedInstanceMap = lo.FilterSliceToMap(instanceTypes, func(it *cloudprovider.InstanceType) (string, *cloudprovider.InstanceType, bool) {
+			return it.Name, it, it != nil
+		})
+		// re-calculate the cost as the instances have changed
+		cost, err := npCost.updateCost()
+		if err != nil {
+			return fmt.Errorf("failed to update cost for nodepool %q: %w", np.Name, err)
+		}
+		cc.npCostMap[np.Name].cost = cost
+	}
+	return nil
+}
+
+func (npc *NodePoolCost) updateCost() (float64, error) {
+	cost := 0.0
+	for ok, oc := range npc.offeringCounts {
+		overlayedInstanceType := npc.overlayedInstanceMap[ok.InstanceName]
+		if overlayedInstanceType == nil {
+			return 0, fmt.Errorf("instance type %q not found in overlayed instance map for offering in zone %q with capacity %q", ok.InstanceName, ok.Zone, ok.Capacity)
+		}
+		// get the offering price from the overlayed instance type
+		newOffering, exists := lo.Find(overlayedInstanceType.Offerings, func(o *cloudprovider.Offering) bool {
+			return o.CapacityType() == ok.Capacity && o.Zone() == ok.Zone
+		})
+		if !exists {
+			return 0, fmt.Errorf("offering not found for instance type %q in zone %q with capacity type %q", ok.InstanceName, ok.Zone, ok.Capacity)
+		}
+		// add the new price times the count of that offering
+		cost = cost + (float64(oc.Count) * newOffering.Price)
+	}
+	return cost, nil
+}
+
+func (cc *ClusterCost) createNewNodePoolCost(np *v1.NodePool, instanceTypes []*cloudprovider.InstanceType) {
+	// create the new npc
+	cc.npCostMap[np.Name] = &NodePoolCost{
+		overlayedInstanceMap: lo.SliceToMap(instanceTypes, func(it *cloudprovider.InstanceType) (string, *cloudprovider.InstanceType) {
+			return it.Name, it
+		}),
+		offeringCounts: make(map[OfferingKey]OfferingCount),
+		cost:           0.0,
+	}
+}
+
+// UpdateNodeClaim adds a NodeClaim to cost tracking. The NodeClaim must have
+// all required labels or it will be ignored and logged as an error.
+func (cc *ClusterCost) UpdateNodeClaim(ctx context.Context, nodeClaim *v1.NodeClaim) {
+	cc.RLock()
+	_, exists := cc.nodeClaimMap[client.ObjectKeyFromObject(nodeClaim).String()]
+	cc.RUnlock()
+	if !exists {
+		// First lets check if the right labels are there
+		var missingLabels []string
+		for _, key := range NecessaryLabels {
+			_, exists := nodeClaim.Labels[key]
+			if !exists {
+				missingLabels = append(missingLabels, key)
+			}
+		}
+		if len(missingLabels) > 0 {
+			log.FromContext(ctx).Error(fmt.Errorf("nodeclaim %q is missing required labels: %v", nodeClaim.Name, missingLabels), "failed to process nodeclaim for cost tracking")
+			CostTrackingErrorsTotal.Inc(map[string]string{
+				metrics.NodePoolLabel: nodeClaim.Labels[v1.NodePoolLabelKey],
+			})
+			return
+		}
+
+		np := &v1.NodePool{}
+		err := cc.client.Get(ctx, client.ObjectKey{Name: nodeClaim.Labels[v1.NodePoolLabelKey]}, np)
```

Comment:
If you use the owner reference you can get the UUID of the owning NodePool and use that for the lookup.